pkg/build/android.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package build
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/google/syzkaller/pkg/osutil"
+)
+
+const (
+	kernelConfig = "common/build.config.gki_kasan.x86_64"
+	moduleConfig = "common-modules/virtual-device/build.config.virtual_device_kasan.x86_64"
+)
+
+type android struct{}
+
+func (a android) runBuild(kernelDir, buildConfig string) error {
+	cmd := osutil.Command("build/build.sh")
+	cmd.Dir = kernelDir
+	cmd.Env = append(cmd.Env, "DIST_DIR=out/dist", fmt.Sprintf("BUILD_CONFIG=%s", buildConfig))
+
+	_, err := osutil.Run(time.Hour, cmd)
+	return err
+}
+
+func (a android) readCompiler(archivePath string) (string, error) {
+	f, err := os.Open(archivePath)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	gr, err := gzip.NewReader(f)
+	if err != nil {
+		return "", err
+	}
+	defer gr.Close()
+
+	tr := tar.NewReader(gr)
+
+	h, err := tr.Next()
+	for ; err == nil; h, err = tr.Next() {
+		if filepath.Base(h.Name) == "compile.h" {
+			bytes, err := ioutil.ReadAll(tr)
+			if err != nil {
+				return "", err
+			}
+			result := linuxCompilerRegexp.FindSubmatch(bytes)
+			if result == nil {
+				return "", fmt.Errorf("include/generated/compile.h does not contain build information")
+			}
+
+			return string(result[1]), nil
+		}
+	}
+
+	return "", fmt.Errorf("archive %s doesn't contain include/generated/compile.h", archivePath)
+}
+
+func (a android) build(params Params) (ImageDetails, error) {
+	var details ImageDetails
+
+	if params.CmdlineFile != "" {
+		return details, fmt.Errorf("cmdline file is not supported for android cuttlefish images")
+	}
+	if params.SysctlFile != "" {
+		return details, fmt.Errorf("sysctl file is not supported for android cuttlefish images")
+	}
+
+	if err := a.runBuild(params.KernelDir, kernelConfig); err != nil {
+		return details, fmt.Errorf("failed to build kernel: %s", err)
+	}
+	if err := a.runBuild(params.KernelDir, moduleConfig); err != nil {
+		return details, fmt.Errorf("failed to build modules: %s", err)
+	}
+
+	buildOutDir := filepath.Join(params.KernelDir, "out", "dist")
+	bzImage := filepath.Join(buildOutDir, "bzImage")
+	vmlinux := filepath.Join(buildOutDir, "vmlinux")
+	initramfs := filepath.Join(buildOutDir, "initramfs.img")
+
+	var err error
+	details.CompilerID, err = a.readCompiler(filepath.Join(buildOutDir, "kernel-headers.tar.gz"))
```

Comment:
Nit: If there's a way to extract the compiler version even in the case of a non-successful build, it would be better to do that - syzbot will include the compiler version also in the `build failed` reports and we don't really want it to be confusing.

Yes, in the existing Linux code there's also a chance that it won't query the compiler version, but since it peeks into the build directory, the build must be totally broken for such thing to happen.

Reply:
I understand the concern.

It's difficult in this case since the non-tarred version of compile.h will be in $ROOT/out/<something>/include/generated/compile.h, where <something> is related to the branch we're building from, but it's difficult to know in advance what it is without passing in some configuration data from the manager config or similar.

We could always walk the out/ tree and look for a compile.h file, but if SyzCI has switched branches and there are still artifacts around from a previous build then we might pick the wrong one. Could this ever happen, or are we safe to just walk the tree and find a <something>/include/generated/compile.h file?